wellcomeml/io/epmc/client.py

```diff
@@ -0,0 +1,302 @@
+            adapter_kwargs = {}
+        session.mount('https://', HTTPAdapter(
+            max_retries=retries, **adapter_kwargs)
+                      )
+        return session
+
+    def _execute_query(self, session, params, only_first=True):
+        """
+        Try to execute the query for the given amount of retries, adding some
+        sleep time every retry.
+        Args:
+            session: A requests session
+            params: A string containing the query
+            only_first: Whether to only return the first result (e.g. in
+            the case of searching for a specific PMID or DOI). Will be
+            deprecated
+        Returns:
+            The response of the query, as a dictionary
+        """
+        if only_first:
+            self.logger.warning("Only returning first entry of search as a "
+                                "dictionary. Will be deprecated in the future"
+                                " and a list of arguments will be"
+                                " returned by default.")
+
+        epmc_search_url = "/".join([self.api_endpoint, "search"])
+
+        go_to_next_page = True
+        page = 1
+
+        final_results = []
+
+        while go_to_next_page:
+            self.logger.debug(f"Querying results from "
+                              f"{page*params['pageSize']} to "
+                              f"{(page+1)*params['pageSize']}")
+
+            response = session.get(
+                epmc_search_url,
+                params=params
+            )
+
+            response_json = response.json()
+
+            current_page_result = response_json["resultList"]["result"]
+
+            if not current_page_result or only_first:
+                go_to_next_page = False
+            else:
+                params['cursorMark'] = response_json.get("nextCursorMark")
+
+            final_results += current_page_result
+            page += 1
+
+        if final_results:
+            return final_results[0] if only_first else final_results
+        else:
+            self.logger.warning(
+                "epmc.license.get_pm_metadata: no-results query=%s", params
+            )
+            return None
+
+    def search(self, session, query, result_type='core', page_size=10,
+               only_first=True, **kwargs):
+        """ Fetches metadata from EPMC's REST API for a query. """
+        # Sample URL for searching by DOI:
+        #   https://www.ebi.ac.uk/europepmc/webservices/rest/search?query=doi:10.1111/hiv.12674
+
+        params = {
+            "query": query,
+            "format": "json",
+            "resultType": result_type,
+            "pageSize": page_size
+        }
+
+        return self._execute_query(session, params, only_first, **kwargs)
+
+    def search_by_doi(self, session, doi, **kwargs):
+        query = "doi:%s" % doi
+        return self.search(session, query, only_first=True, **kwargs)
+
+    def search_by_pmid(self, session, pmid, **kwargs):
```

Comment:
`search_by_pmid`, `search_by_pmids` and `search_by_doi`, `search_by_dois` are not tested either (mea culpa!). If you want to add tests for those, that'd be really nice, otherwise, let's open a ticket and work on it later.

Reply:
all of those are pretty thin wrappers around search. they are quite easy to test, I will just patch search and make sure it is called with the right args

Reply:
7c3b2293a6fa061eb878e5b78debd81b27b51a6f